### DISCUSSION TITLE
How to translate this 2p-specific code in 2p-kt?

### DISCUSSION THREAD

**Discussion by enrdenti** [tuProlog/2p-kt#16] [Q&A] [answered]
Issue: How to convert old-fashioned Java code for classic 2p into newer Java (perhaps Kotlin) code for 2p-kt

What is needed: the equivalent of basic methods to instantiate a prolog engine and use it from Java (kotlin), such as in the following snapshot:

```java
Prolog engine = new Prolog(); 
Theory t = new Theory(new FileInputStream("exp_parser.pl"));
engine.setTheory(t);
SolveInfo info = engine.solve("evalExpr(" + prologInputString + ",Result).");
if (info.isSuccess()) { Term resultTerm = info.getTerm("Result");  …}
```

**Comment by gciatto** [comment, marked as answer]
Hello @enrdenti.

[Here](https://github.com/tuProlog/2pkt-java-example/blob/master/src/main/java/it/unibo/tuprolog/java/example/Example.java)'s a complete rewrite of your code.

There are a couple of things to be taken into account:

1. you need to have at least modules `:solve-classic` and `:parser-theory` (as well as their direct and indirect dependencies) in your classpath, as you need both resolution facilities (term representation + unification + theories + resolution) and parsing facilities (for both terms and theories)
    - hence, if you manage dependencies with Gradle (like I do [here](https://github.com/tuProlog/2pkt-java-example/blob/master/build.gradle.kts) you may be willing to add the following section:
    ```kotlin
    dependecies {
        implementation("it.unibo.tuprolog:solve-classic-jvm:$tuprologVersion")
        implementation("it.unibo.tuprolog:parser-theory-jvm:$tuprologVersion")
    }
    ```

2. you have to keep in mind the following conceptual map:
    - There is no `Prolog` class anymore, there's a new unified interface for all solvers: `Solver`
    - `Solver`s are created via some `SolverFactory`
    - There may exist several sorts of `Solver`s, hence as many sorts of `SolverFactory`s
        * the "classic" solver is the one working similarly to 2P's `Prolog` engines 
    - In any case solvers can be either `Solver`s (__unmodifiable__ interface) or `MutableSolver`s (__modifiable__ interface)
        * unmodifiable solvers (i.e. instances of`Solver`) expose no method to vary their internal status (e.g. dynamic KB, flags, etc.), yet such status may still vary as a by-product of Prolog queries being executed
        * modifiable solvers (i.e. instances of`MutableSolver`) expose many methods to vary their internal status (e.g. dynamic KB, flags, etc.) from Java
        * notice that `MutableSolver` is a __sub-type__ of `Solver`
    - To instantiate a `Solver` you must instantiate a `SolverFactory` first
        * you can instantiate a `SolverFactory` for "classic" `Solver`s via either `Solve.getClassic` or `ClassicSolverFactory.INSTANCE`
    - Each instance of `SolverFactory` comes with several methods to instantiate `Solver`s with, namely:
        * `factory.`__`solverOf`__`(...)` creating an __empty__ (i.e. with no built-in) and __unmodifiable__ `Solver` instance
        * `factory.`__`solverWithDefaultBuiltins`__`(...)` creating an __unmodifiable__ `Solver` instance with default Prolog builtins
        * `factory.`__`mutableSolverOf`__`(...)` creating an __empty__ (i.e. with no built-in) and __modifiable__  instance of `MutableSolver`
        * `factory.`__`mutableSolverWithDefaultBuiltins`__`(...)` creating an instance of `MutableSolver` with default Prolog builtins
        * notice that all such methods come with several overloads letting the caller specify initialization aspects of the to-be-created solvers (e.g. static KB, flags, input/output/warning channels etc.)
    - So, if you create a `Solver`, the `Theory` must be provided as an argument upon creation, whereas if you create a `MutableSolver` the static KB can be edited even after the solver has been instantiated
        * hence, there is no equivalent of `.setTheory(...)` on `Solver`s, while there is `.loadStaticKB(...)` on `MutableSolver`s 

3. concerning parsing, keep in mind that
    -`Theory`s and `Terms` are just __immutable__ data structures
    - They can be either _parsed_ from a string or _read_ a file via the following classes:
        * __`TermParser`__ whose instances aim at parsing single terms (or structs, or whatever) from in-memory strings
        * __`TermReader`__ whose instances aim at reading single terms (or structs, or whatever) from input streams or readers (which may be files, sockets, or in-memory buffers)
        * __`ClausesParser`__ whose instances aim at parsing sequences of clauses (or theories) from in-memory strings, either lazily or eagerly
        * __`TermReader`__ whose instances aim at parsing sequences of clauses (or theories) from input streams or readers (which may be files, sockets, or in-memory buffers), either lazily or eagerly
    - Notice that all such types must be constructed out of an `OperatorSet`, which specifies which symbols should be considered as operators while parsing (and with which associativity and priority)
        * static factory methods are available to create readers/parsers out of default Prolog operators

4. notice that, because of the single responsibility principle, it is not `Solver`s' responsibility to perform parsing of queries
   - hence there is no `solve` method accepting a string as query
   - the query `Struct` must be parsed before calling the solver's `solve` method, via some `TermParser`

5. notice that there is no such a notion like 2P's `SolveInfo`: `Solver`s' `solve` method returns a [`Sequence`](https://kotlinlang.org/docs/sequences.html) of `Solution`s
   - `Solution` is an interface, and it is the base type for all logic solutions
       * it comes with three subtypes: `Solution.Yes` for successful answers, `Solution.No` for negative answers, and `Solution.Halt` for exceptions
       * by testing the type of a `Solution` via `instanceof` one may figure out what situation is at hand
       * by casting a `Solution` to the most adequate subtype, one may access relevant information, such as the resolving `Substitution` (from which variable bindings can be retrieved) of a `Solution.Yes` or the exception which caused a `Solution.Halt`
   - `Sequences<T>` are (possibly lazy) streams of data of type `T`
       * they can be conveniently handled in Kotlin via extension methods
       * in Java they could be converted into instances of `Stream` (cf. <https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.streams/kotlin.sequences.-sequence/as-stream.html>)...
       * ... or iterated via the iterator pattern:
       ```java
        Iterator<Solution> si = solver.solve(….).iterator();
        while (si.hasNext() {
            Solution solution = si.next();
            // handle solution here
        }
       ```

6. notice that there is a convenient abstraction: `SolutionFormatter` which aims at converting solutions into presentable strings
   - hence if you plan to show your solver's solutions to the user you may consider implementing your own `SolutionFormatter` or use some default implementation instead

**Reply by enrdenti** [reply]
Thanks a lot! One question: since my example theory needs to introduce a custom operator, namely `op(400,yfx,':')`, how should I get a customised `TermParser` that includes it? (actually, putting an `:-op` directive inside the `.pl` source file does not make the job). That is, which method call should I use instead of `TermParser termParser = TermParser.getWithDefaultOperators()` ? I checked the online documentation but there seems to be no info on `TermParser`'s companion object.

Then, one curiosity: why does `ClausesReader.readTheory` expects an `InputStream` rather than a `Reader`?
Thanks!

**Reply by gciatto** [reply]
Have you already tried to add an `op` directive to the theory file? If I well remember it should work.

Indeed, if a parsed `Theory` contains a custom operator directive, and that theory is loaded into a `Solver`, then that solver's `getOperators()` method should return an `OpetarorSet` including the custom operators defined into the theory. 

However, to answer your question, you must create/obtain a custom `OperatorSet` first, then pass it to the `TermParser.withOperators` factory method.
Otherwise, you can use a some solver's `getOperators()` method to retrieve the exact operator set used by that solver. 

A new operator set extending Prolog's standard operators can be created as follows:
```java
OperatorSet customOpSet = OperatorSet.STANDARD.plus(new Operator(":", Specifier.YFX, 400))
```

Thus, a term parser with custom operators can be created via
```java
TermParser.withOperators(customOpSet)
```
or via 
```java
TermParser.withOperators(solver.getOperators())
```

**Reply by gciatto** [reply]
`ClausesReader.readTheory` should come with an overload accepting `Reader`s as well, if I well remember.

If it doesn't, then please open an issue since such an overload should definitely exist.

**Reply by gciatto** [reply]
I've checked the code.
`ClausesReader.readTheory` does indeed come with an overload accepting `Reader`s: https://github.com/tuProlog/2p-kt/blob/7ac4a41e5ee41f49e443ac07528956567b70e290/parser-theory/src/jvmMain/kotlin/it/unibo/tuprolog/theory/parsing/ClausesReader.kt#L23

**Reply by enrdenti** [reply]
> Have you already tried to add an `op` directive to the theory file? If I well remember it should work.

Yes, it is the first thing I tried, since it did work on classic 2p, but it is actually ignored - don't know if on purpose or due to a bug.
Note that, instead, both 2p-kt-IDE and 2p-kt-repl do recognise and treat it correctly.

**Reply by enrdenti** [reply]
Hello @gciatto-unibo,

[Here (HybridParser.java.txt)](https://github.com/tuProlog/2p-kt/files/7400124/HybridParser.java.txt) is the modified version that actually handles my custom operator.

The key changes are as follows:
```java
// -- commented out by ED to bypass the lack of support to reading :-op directives from file
// ClausesReader theoryReader = ClausesReader.getWithDefaultOperators();
// TermParser termParser = TermParser.getWithDefaultOperators();
// -- added by ED to add support for extra operator
OperatorSet customOpSet = OperatorSet.STANDARD.plus(new Operator(":", Specifier.YFX, 400));
TermParser termParser = TermParser.withOperators(customOpSet);
ClausesReader theoryReader = ClausesReader.withOperators(customOpSet);
// -- ED ends
```
Suggestion: in the online documentation, please addthe class full absolute name (i.e. with package specification) to each class documentation, otherwise it is quite hard to browse the package list to find the appropriate import to add. Thanks!

**Reply by gciatto** [reply]
> Yes, it is the first thing I tried, since it did work on classic 2p, but it is actually ignored - don't know if on purpose or due to a bug.
Note that, instead, both 2p-kt-IDE and 2p-kt-repl do recognise and treat it correctly.

Could you please be more precise? What is ignored exactly?
While reading/parsing a theory, if the Prolog code to be read/parsed contains an `:- op/3` directive, the corresponding operator is added to the operator set used for parsing.
If a `Theory` containing some `:- op/3` directive is loaded into a `Solver`, then each operator defined in it is be automatically loaded into that `Solver`, as an `Operator`.

What is deliberately non-automatic is that any new `TermParser` created after the `Solver` will not inherit any operator from the `Solver`, unless explicitly stated via `TermParser.withOperators(solver.getOperators())`.
In fact, why should it? `TermParser`s and `Solvers` are different objects, and they do not share any data.
`OperatorSet` are __immutable__ containers of `Operator`s.
So, even if you create a `TermParser` out of a `Solver` via  `TermParser.withOperators(solver.getOperators())`, and you edit the pool of `Operator`s loaded into the `Solver`, the pool of operators loaded into the `TermParser` will __not__ change.

**Reply by gciatto** [reply]
> Here (HybridParser.java.txt) is the modified version that actually handles my custom operator.

This is fine, but [here](https://github.com/tuProlog/2pkt-java-example/blob/master/src/main/java/it/unibo/tuprolog/java/example/Example.java)'s a "more correct" version of the code which explicitly does what (I believe) you intended.
- notice that I have added a directive to the [theory](https://github.com/tuProlog/2pkt-java-example/blob/master/src/main/resources/it/unibo/tuprolog/java/example/ExampleTheory.pl) I have used in my example: as you can see, `op/3` directives are indeed automatically handled

**Reply by gciatto** [reply]
> Suggestion: in the online documentation, please addthe class full absolute name (i.e. with package specification) to each class documentation, otherwise it is quite hard to browse the package list to find the appropriate import to add. Thanks!

Wish I had enough room to do that... 

Counter-suggestion: if you use some IDE like Eclipse or IntelliJ Idea to perform coding experiments, automatic import suggestions should be supported (e.g. Ctrl+Space on Eclipse or Alt+Enter on IntelliJ)

**Reply by enrdenti** [reply]
Perfect, this is exactly what I meant to do, thanks a lot!
Just two final minor comments: 
- in the line `SolverFactory solverFactory = ClassicSolverFactory.INSTANCE;` you suggest the possible alternative `// or Solver.getClassic()`: I tried it but found it somehow misleading, as an idiom such as `Solver.getClassic()`  leads to think that the result is a `Solver`, not a `SolverFactory`. So, the first alternative is far more clear. Yet, the second could be viable, effective and intuitive if moved into the `SolverFactory` class.
- as concerns the operator handling: I do appreciate that operator directives are handled automatically in solvers, and also appreciate that `TermParser` is both conceptually and physically separate entity. However, putting the two aspects together, usability can become somehow counter-intuitive, in that if the user means to parse terms (typically, queries) to be submitted to a solver that includes some user-defined operators, having to re-define such operators in the `TermParser` is weird. Moreover, for the solver to work, I guess it internally does instantiate and use a `TermParser` with the suitable operator set already in place. So, why not just make **that** `TermParser` accessible, providing a method such as `solver.getTermParser`? That would be straightforward (and also avoid duplicating the parser object): one would just re-use the inherently correct term parser, borrowed from the solver.

**Reply by gciatto** [reply]
> In the line `SolverFactory solverFactory = ClassicSolverFactory.INSTANCE;` you suggest the possible alternative `// or Solver.getClassic()`: I tried it but found it somehow misleading, as an idiom such as `Solver.getClassic()` leads to think that the result is a `Solver`, not a `SolverFactory`. 

I agree on your concern. Indeed, while the current idiom is a by-product of another idiom conceived for Kotlin.
The original intention, in Kotlin, was to support the following notation:
```kotlin
Solver.classic.solverOf(....)
// or
Solver.classic.solverWithDefaultBuiltins()
```
The `getClassic()` idiom "occurs" because Kotlin properties are compiled into Java getters.
However, I agree that better alternatives should be made more intuitive.
I've opened an issue on GitLab to keep track of this topic: <https://gitlab.com/pika-lab/tuprolog/2p-in-kotlin/-/issues/157>.

**Reply by gciatto** [reply]
> So, the first alternative is far more clear. Yet, the second could be viable, effective and intuitive if moved into the `SolverFactory` class.

I disagree. It is a common practice in JVM-based languages, when one has the `X`, `XFactory` and/or `XBuilder` types to allow the instantiation of factories or builders from the "main" type -- `X`, in my example -- via ad-hoc static methods.

**Reply by gciatto** [reply]
> As concerns the operator handling: I do appreciate that operator directives are handled automatically in solvers, and also appreciate that `TermParser` is both conceptually and physically separate entity. However, putting the two aspects together, usability can become somehow counter-intuitive, in that if the user means to parse terms (typically, queries) to be submitted to a solver that includes some user-defined operators, having to re-define such operators in the `TermParser` is weird. Moreover, for the solver to work, I guess it internally does instantiate and use a `TermParser` with the suitable operator set already in place. So, why not just make **that** `TermParser` accessible, providing a method such as `solver.getTermParser`? That would be straightforward (and also avoid duplicating the parser object): one would just re-use the inherently correct term parser, borrowed from the solver. 

No, your guess is not correct, this time :)
`Solver`s hide no parser behind the scenes, and this is not only intended but strictly enforced.
The rationale behind this choice is, once again, the single responsibility principle: `Solver`s are not in charge of parsing terms, nor theories. 
Indeed, one can only either create a `Solver` out of an already-parsed `Theory`, or load an already-parsed `Theory` into an already created `MutableSolver`. In both cases, the `Theory` **must** be parsed/read _externally_ w.r.t. the `Solver`.

More generally, if you look carefully at the [project map](https://github.com/tuProlog/2p-kt/blob/master/.img/project-map.png), you will notice that there is no direct dependency among any `:solver-*` and any `:parser-*` module. 
This is deliberate.
The goal is to make it difficult for the people developing 2P-Kt (or some new module in the ecosystem) to abuse the exploitation of the parser, unless strictly necessary.
This, in turn, is aimed at avoiding core functionalities (e.g. resolution, and the Prolog standard builtins) to use the parser behind the scenes, which would imply potential inefficiencies.
(I really wanted to avoid some errors which I found in the implementation of 2P-Java, a few years ago, such as cloning terms by converting them into strings and parsing them back)

Of course, there exist standard IO builtins that technically require parsing, such as `consult/1`, or `read*/N`.
These are indeed placed into the `io-lib` module which does rely on the parser behind the scenes.

So, long story short, if you need a quick way to create a `TermParser` out of a `Solver`, you can simply create your own utility method, such as:
```java
public static TermParser termParserFor(Solver solver) {
    return TermParser.withOperators(solver.getOperators);
}

public static Struct parseGoal(Solver solver, String string) {
    return termParserFor(solver).parseStruct(string);
}

public static Sequence<Solution> solve(Solver solver, String query, SolveOptions options) {
    return solver.solve(parseGoal(query), options);
}

public static Sequence<Solution> solve(Solver solver, String query) {
    return solve(solver, query, SolveOptions.DEFAULT);
}
``` 

This solution would of course imply the creation of several `TermParser`s... yet don't worry: `TermParser`s are very lightweight objects, and they can be created very quickly.

**Reply by enrdenti** [reply]
I possibly mis-explained my point, sorry. The fact is that solvers do interpret the :-op/3 directives, as you mentioned; and they do so by loading a theory, which - then - must have already been parsed. Your example actually does so via the theoryReader object, which is an instance of ClausesReader (with default operators, btw). So, I possibly spoke about "term parsers" when should have said "clauses reader", but as a matter of fact it seems someone there does parse :-op/3 and the theory, which is then passed to the solver (and I do agree with the separation). Correct?

**Reply by gciatto** [reply]
Yes, you're correct.

There are 2 disjoint, yet interrelated, mechanisms here:
1. `ClausesReader`s and `ClausesParser`s can consider `:- op/3` directives while reading/parsing, adding the corresponding operators to the set of operators used for reading/parsing the remainder of the file. The result is usually a theory containing a `Directive` object, which is a particular case of `Clause`
2. `Solver`s have their own set of operators as part of their internal state. This is accessible via `Solver.getOperators()`. It may evolve because of queries being solved. However, a `Solver`'s operator set is initialised by merging: 
    - all operators explicitly declared by each library loaded into that `Solver` (in the case of Prolog standard builtin libraries, this means the whole set of Prolog standard operators)
    - all operators implicitly declared in the `Solver`'s static and dynamic KBs (which are `Theory`s), as `:- op/3` directives

**Reply by enrdenti** [reply]
perfect! crystal clear now
